crates/circuit/src/circuit_data.rs

```diff
@@ -1957,6 +2535,172 @@ impl CircuitData {
             _ => self.set_global_phase(add_global_phase(&self.global_phase, value)?),
         }
     }
+
+    /// Add a classical variable to the circuit.
+    ///
+    /// # Arguments:
+    ///
+    /// * var: the new variable to add.
+    /// * type_: the type the variable should have in the circuit.
+    ///
+    /// # Returns:
+    ///
+    /// The [Var] index of the variable in the circuit.
+    pub fn add_var(&mut self, var: expr::Var, type_: CircuitVarType) -> PyResult<Var> {
+        let name = {
+            let expr::Var::Standalone { name, .. } = &var else {
+                return Err(CircuitError::new_err(
+                    "cannot add variables that wrap `Clbit` or `ClassicalRegister` instances",
+                ));
+            };
+            name.clone()
+        };
+
+        match self.identifier_info.get(&name) {
+            Some(CircuitIdentifierInfo::Var(info)) if Some(&var) == self.vars.get(info.var) => {
+                return Err(CircuitError::new_err("already present in the circuit"));
+            }
+            Some(_) => {
+                return Err(CircuitError::new_err(
+                    "cannot add var as its name shadows an existing identifier",
+                ));
+            }
+            _ => {}
+        }
+
+        match type_ {
+            CircuitVarType::Input
+                if !self.vars_capture.is_empty() || !self.stretches_capture.is_empty() =>
+            {
+                return Err(CircuitError::new_err(
+                    "circuits to be enclosed with captures cannot have input variables",
+                ));
+            }
+            CircuitVarType::Capture if !self.vars_input.is_empty() => {
+                return Err(CircuitError::new_err(
+                    "circuits with input variables cannot be enclosed, so cannot be closures",
```

Comment:
The writing on this error message is a bit strange. I tried adding an article to make it more understandable but it might need to be rewritten further.
```suggestion
                    "circuits with input variables cannot be enclosed, so they cannot be closures",
```

Reply:
Checkpoint (Continue reviewing from here...)